airflow/lineage/__init__.py

```diff
@@ -133,13 +133,20 @@ def wrapper(self, context, *args, **kwargs):
 
             # Remove auto and task_ids
             self.inlets = [i for i in self.inlets if not isinstance(i, str)]
-            _inlets = self.xcom_pull(context, task_ids=task_ids, dag_id=self.dag_id, key=PIPELINE_OUTLETS)
-
-            # re-instantiate the obtained inlets
-            # xcom_pull returns a list of items for each given task_id
-            _inlets = [item for item in itertools.chain.from_iterable(_inlets)]
-
-            self.inlets.extend(_inlets)
+            # We manually create a session here since xcom_pull returns a LazyXComAccess iterator.
+            # If we do not pass a session a new session will be created, however that session will not be
+            # properly closed and will remain open. After we are done iterating we can safely close this
+            # session.
+            with create_session() as session:
+                _inlets = self.xcom_pull(
+                    context, task_ids=task_ids, dag_id=self.dag_id, key=PIPELINE_OUTLETS, session=session
+                )
```

Comment:
```suggestion
             _inlets = self.xcom_pull(
                    context, task_ids=task_ids, dag_id=self.dag_id, key=PIPELINE_OUTLETS
                )
```
The call to `create_session` is no longer needed since you decorated `xcom_pull` with `@provide_session`. The session will be automatically provided

Reply:
Am i missing something @uranusjr ?

Reply:
I think both work (and using the default session from `@provide_session` is cleaner so you’re correct)

Reply:
Removing the create_session will actually remove the bug fix. As explained in the comment above the create session: the issue is that xcom_pull returns a LazyXComAccess iterator. Since we have returned from the xcom_pull function, the session that would have been automatically created there will be closed and thus a new one that leaks will be started when we iterate. By wrapping this in a manually created session the leak is fixed

Reply:
Ah I see, the point is to make the session live slightly longer so line 147 below can still use it.

Reply:
Maybe a comment above this block would help future readers understand the context better. 

(Also I think the blank line + comments in lines 144-146 might actually hurt readability a bit since they make the two lines feel less associated. Maybe moving them above the entire block would help.)

Reply:
I remove the empty lines and 2 comments. I don't think the added much here.

There already is a comment above the create_session block to explain things, so I am not sure I correctly understand what you mean with `Maybe a comment above this block would help future readers understand the context better.`. Can you clarify it for me?